src/commands/deploy/deploy.ts

```diff
@@ -54,7 +55,17 @@ export async function deploy(actionContext: IActionContext, arg1?: vscode.Uri |
     // if we already got siteConfig, don't waste time getting it again
     siteConfig = siteConfig ? siteConfig : await node.root.client.getSiteConfig();
 
-    if (javaUtils.isJavaRuntime(siteConfig.linuxFxVersion)) {
+    let javaMavenModule = undefined;
+    if (siteConfig && javaUtils.isJavaRuntime(siteConfig)) {
+        fileExtensions = fileExtensions || await javaUtils.getJavaFileExtensions(siteConfig);
+        javaMavenModule = javaUtils.getMavenModule(context.effectiveDeployFsPath);
+        // get packaging type as artifact file extension for maven module.
+        const ext = javaMavenModule?.packaging ?? context.effectiveDeployFsPath.split('.').pop() ?? '';
+        if (!fileExtensions?.includes(ext)) { // show error and break if deploying a raw folder/or non-matching artifact to java runtime instance.
+            void vscode.window.showErrorMessage(localize('errorArtifactsNotMatchingJava', 'Only "{0}" file/maven module can be deployed to "{1}"', `${fileExtensions}`, siteConfig.name));
```

Comment:
We don't want to block users from deploying like this. Once they've added settings like "preDeployTask" and/or "deploySubpath", they should have full control and we will just do what they specified in the settings.

Reply:
OK, I will change this part to allow proceeding and show a warning instead of error.

but I doubt if it's a good UX to allow proceeding when we know it will certainly fail at last.